### PR TITLE
add setting locale method via setting app

### PR DESCRIPTION
**Pull request by KazuCocoa**
We can set the locale via [the setting app feature](https://github.com/appium/io.appium.settings/pull/11) after the latest the app released without rebooting devices after changing the locale.

After merging this and update `helpers.ensureDeviceLocale` in `appium-android-driver`, we'll be able to set locale for Android N+ preventing https://github.com/appium/appium/issues/9619  issue.

-----

Update

- ~~I've added some `skip` to avoid travis failed which also have been happened on current master~~ => https://github.com/appium/appium-adb/pull/272
    - https://github.com/appium/appium-adb/pull/271#issuecomment-346260317
    - https://github.com/appium/appium-adb/pull/271/commits/c8d31a57944a61a41347cddffa1cea5773b95e16
    - I'll create an issue to fix it. But in this PR, I add `skip` to focus on current changes.

After merging this PR, I'll update:

- https://github.com/appium/appium-android-driver/blob/ed07b52eab38865da619877ba90eb068d758964f/lib/android-helpers.js#L91
    - remove existing duplicated logic and update test cases
    - update package.json to use the latest `appium-adb` and `io.appium.settings`

### CODE REVIEW

File: test/functional/adb-commands-e2e-specs.js [hunk 1]
```diff
@@ -151,6 +151,17 @@ describe('adb commands', function () {
     await adb.setAnimationState(true);
     (await adb.isAnimationOn()).should.be.true;
   });
+  it('should be able to set device locale via setting app @skip-ci', async () => {
+    // Operation not allowed: java.lang.SecurityException: Package io.appium.settings has not requested permission android.permission.CHANGE_CONFIGURATION
+    // is shown if the setting apk is not updated.
+    await adb.grantPermission('io.appium.settings', 'android.permission.CHANGE_CONFIGURATION');
+
+    await adb.setDeviceSysLocaleViaSettingApp('fr', 'fr');
+    (await adb.getDeviceSysLocale().should.equal('fr-FR'));
```
Comment (mykola-mokhnach) [inline test/functional/adb-commands-e2e-specs.js]: I assume only `await adb.getDeviceSysLocale()` should be enclosed into parentheses. Same thing below.

File: lib/tools/adb-commands.js [hunk 2]
```diff
@@ -770,6 +772,22 @@ methods.isAnimationOn = async function () {
...
+/**
+ * Change the locale on the device under test. Don't need to reboot the device after changing the locale.
+ * This method sets an arbitrary locale following:
+ *   https://developer.android.com/reference/java/util/Locale.html
+ *   https://developer.android.com/reference/java/util/Locale.html#Locale(java.lang.String,%20java.lang.String)
+ *
+ * @param {string} language - Language. e.g. en, ja
+ * @param {string} country - Country. e.g. US, JP
+ */
+methods.setDeviceSysLocaleViaSettingApp = async function (language, country) {
```
Comment (mykola-mokhnach) [inline lib/tools/adb-commands.js]: I was thinking about some approach, which would allow us to isolate country/language selection logic on appium-adb module level instead of changing it on driver level. For example we call something like `adb.setDeviceLocale` and the internal logic automatically decides which method selection to apply based on API level, so we don't have to tune that for each driver.
How do you think @imurchie ?

Comment (KazuCocoa) [reply lib/tools/adb-commands.js]: 👍 I'm also thinking so (the driver just call `adb.setDeviceLocale` for example) when I read codebase and added this method.

Comment (mykola-mokhnach) [reply lib/tools/adb-commands.js]: you can implement this approach, so then we could increase major module version and update client drivers when it is published

File: lib/tools/apk-utils.js [hunk 3]
```diff
@@ -474,12 +474,81 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
+    if (locale !== curLocale) {
+      await this.setDeviceSysLocale(locale);
+      changed = true;
+    }
+  } else { // API >= 24
+    let curLocale = await this.getDeviceLocale();
+    let localeLang;
+    let localeCountry;
+    if (!haveCountry) {
+      localeLang = language.toLowerCase();
```
Comment (KazuCocoa) [inline lib/tools/apk-utils.js]: Can cover `fr` case, for example.

File: lib/tools/apk-utils.js [hunk 4]
```diff
@@ -474,12 +474,81 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
+apkUtilsMethods.setDeviceLanguageCountry = async function (language, country) {
+  let haveLanguage = language && typeof language === "string";
+  let haveCountry = country && typeof country === "string";
+  if (!haveLanguage && !haveCountry) {
+    return;
+  }
+  let changed = false;
+  let apiLevel = await this.getApiLevel();
+
+  if (apiLevel < 23) {
```
Comment (KazuCocoa) [inline lib/tools/apk-utils.js]: `< 24` cases are almost moved from https://github.com/appium/appium-android-driver/blob/ed07b52eab38865da619877ba90eb068d758964f/lib/android-helpers.js#L91

**Comment by KazuCocoa** [comment]
Same error on current master. (Travis)
https://travis-ci.org/appium/appium-adb/jobs/295270516

I'll take a look more.

File: lib/tools/apk-utils.js [hunk 5]
```diff
@@ -474,12 +474,81 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
+  await this.setDeviceLanguageCountry(split_locale[0], split_locale[1]);
+};
+
+/**
+ * Set the locale name of the device under test.
+ *
+ * @param {string} language - Language. e.g. en, ja
+ * @param {string} country - Country. e.g. US, JP
+ */
+apkUtilsMethods.setDeviceLanguageCountry = async function (language, country) {
```
Comment (KazuCocoa) [inline lib/tools/apk-utils.js]: - in `lib/tools/adb-commands.js`, focus on calling adb commands
- in `lib/tools/apk-utils.js`, handle adb commands

File: lib/tools/apk-utils.js [hunk 6]
```diff
@@ -474,12 +474,81 @@ apkUtilsMethods.getDeviceLocale = async function () {
 
 /**
  * Set the locale name of the device under test.
+ * Compatibility for appium-android-driver 1.32.4-
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: please do not add explicit driver versions, since not only android driver depends on this module

File: lib/tools/apk-utils.js [hunk 7]
```diff
@@ -474,12 +474,81 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
  *
  * @param {string} locale - The name of the new device country.
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: please add an example value(s)

File: lib/tools/apk-utils.js [hunk 8]
```diff
@@ -474,12 +474,81 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
+};
+
+/**
+ * Set the locale name of the device under test.
+ *
+ * @param {string} language - Language. e.g. en, ja
+ * @param {string} country - Country. e.g. US, JP
+ */
+apkUtilsMethods.setDeviceLanguageCountry = async function (language, country) {
+  let haveLanguage = language && typeof language === "string";
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: use `_.isString`

File: lib/tools/apk-utils.js [hunk 9]
```diff
@@ -474,12 +474,81 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
+  let haveCountry = country && typeof country === "string";
+  if (!haveLanguage && !haveCountry) {
+    return;
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: please add some logging here

File: lib/tools/apk-utils.js (hunk 8 above)
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: -> hasLanguage

File: lib/tools/apk-utils.js [hunk 10]
```diff
@@ -474,12 +474,81 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
+
+/**
+ * Set the locale name of the device under test.
+ *
+ * @param {string} language - Language. e.g. en, ja
+ * @param {string} country - Country. e.g. US, JP
+ */
+apkUtilsMethods.setDeviceLanguageCountry = async function (language, country) {
+  let haveLanguage = language && typeof language === "string";
+  let haveCountry = country && typeof country === "string";
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: -> hasCountry

File: lib/tools/apk-utils.js [hunk 11]
```diff
@@ -474,12 +474,81 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
+  if (!haveLanguage && !haveCountry) {
+    return;
+  }
+  let changed = false;
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: -> wasSettingChanged

File: lib/tools/apk-utils.js [hunk 12]
```diff
@@ -474,12 +474,81 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
+    let curCountry = await this.getDeviceCountry();
+    if (haveLanguage && language !== curLanguage) {
+      await this.setDeviceLanguage(language);
+      changed = true;
+    }
+    if (haveCountry && country !== curCountry) {
+      await this.setDeviceCountry(country);
+      changed = true;
+    }
+  } else if (apiLevel < 24) { // API == 23
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: why not to say `apiLevel === 23` ? ;)

File: lib/tools/apk-utils.js [hunk 13]
```diff
@@ -474,12 +474,81 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
+    let curLocale = await this.getDeviceLocale();
+    let locale;
+    if (!haveCountry) {
+      locale = language.toLowerCase();
+    } else if (!haveLanguage) {
+      locale = country;
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: country.toUppperCase?

File: lib/tools/apk-utils.js [hunk 14]
```diff
@@ -474,12 +474,81 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
+    log.debug(`Current locale: '${curLocale}'; requested locale: '${locale}'`);
+    if (locale !== curLocale) {
+      await this.setDeviceSysLocale(locale);
+      changed = true;
+    }
+  } else { // API >= 24
+    let curLocale = await this.getDeviceLocale();
+    let localeLang;
+    let localeCountry;
+    if (!haveCountry) {
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: Can we refactor the code to avoid duplicating this condition for api == 23/api > 23 cases?

File: test/unit/logcat-specs.js [hunk 15]
```diff
@@ -12,7 +12,8 @@ describe('logcat', async () => {
   let adb = {path: 'dummyPath', defaultArgs: []};
   let logcat = new Logcat({adb, debug: false, debugTrace: false});
   describe('startCapture', withMocks({teen_process}, (mocks) => {
-    it('should correctly call subprocess and should resolve promise', async () => {
+    // Because of  ExpectationError: Unexpected call: resolve(...) on Travis
```
Comment (mykola-mokhnach) [inline test/unit/logcat-specs.js]: Shouldn't it be a separate PR?

Comment (KazuCocoa) [reply test/unit/logcat-specs.js]: ah, yes. I'll create another PR to skip them and rebase after this PR after merge it.

File: lib/tools/apk-utils.js [hunk 16]
```diff
@@ -474,12 +474,76 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
-  await this.setDeviceSysLocale(locale);
+  let split_locale = locale.split("-");
+  await this.setDeviceLanguageCountry(split_locale[0], split_locale[1]);
+};
+
+/**
+ * Set the locale name of the device under test.
+ *
+ * @param {string} language - Language. format: [a-zA-Z]{2,8}. e.g. en, ja : https://developer.android.com/reference/java/util/Locale.html
+ * @param {string} country - Country. format: [a-zA-Z]{2} | [0-9]{3}. e.g. US, JP : https://developer.android.com/reference/java/util/Locale.html
```
Comment (KazuCocoa) [inline lib/tools/apk-utils.js]: I've changed don't convert the language and the country to _upper case_ or _lower case_ in `setDeviceLanguageCountry` because they should accept `[a-zA-Z]` as Locale spec.

File: lib/tools/apk-utils.js (hunk 14 above)
Comment (KazuCocoa) [reply lib/tools/apk-utils.js]: BTW, we can **remove** the `api == 23` case and use `setDeviceSysLocaleViaSettingApp` instead since I've used `setDeviceSysLocaleViaSettingApp` logic to change device locale for all devices in my company and the devices include api 23.

^ is just my thinking and what do you think @mykola-mokhnach ? (It's okay if you have no exact opinion, though)
But I also think it's better to keep current behaviour, using adb command, for the `api == 23/api > 23` since they work fine so far. 🤔

File: lib/tools/apk-utils.js [hunk 17]
```diff
@@ -474,12 +474,76 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
+
+    log.debug(`Current locale: '${curLocale}'; requested locale: '${locale}'`);
+    if (locale !== curLocale) {
+      await this.setDeviceSysLocale(locale);
+      wasSettingChanged = true;
+    }
+  } else { // API >= 24
+    let curLocale = await this.getDeviceLocale();
+    if (!hasCountry || !hasLanguage) {
+      log.debug(`setDeviceLanguageCountry requires language and country for API 24+`);
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: -> ... requires both language and country to be set for ...

File: lib/tools/apk-utils.js [hunk 18]
```diff
@@ -474,12 +474,76 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
+      log.debug(`language: '${language}' and country: '${country}'`);
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: ... Got language: ...

File: lib/tools/apk-utils.js (hunk 14 above)
Comment (mykola-mokhnach) [reply lib/tools/apk-utils.js]: The logic is ok there, no need to remove anything. I just wanted to optimise some redundant code and the recent state of this code block makes me happy :)

**Comment by KazuCocoa** [comment]
✅  and updated.
After merging https://github.com/appium/appium-adb/pull/272 and rebase to the latest master, I'll CI should green.

File: lib/tools/apk-utils.js [hunk 19]
```diff
@@ -473,13 +473,79 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
+      }
+
+      log.debug(`Current locale: '${curLocale}'; requested locale: '${locale}'`);
+      if (locale !== curLocale) {
+        await this.setDeviceSysLocale(locale);
+        wasSettingChanged = true;
+      }
+    } else { // API >= 24
+      if (!hasCountry || !hasLanguage) {
+        log.debug(`setDeviceLanguageCountry requires both language and country to be set for API 24+`);
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: I'd rather change the level to `warn` and add a mention about device locale is not going to be changed

File: lib/tools/apk-utils.js [hunk 20]
```diff
@@ -473,13 +473,79 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
+
+      log.debug(`Current locale: '${curLocale}'; requested locale: '${language}-${country}'`);
+      if (`${language}-${country}` !== curLocale) {
+        await this.setDeviceSysLocaleViaSettingApp(language, country);
+      }
+    }
+  }
+
+  if (wasSettingChanged) {
+    await this.reboot();
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: info log message here might be useful to explain why reboot is needed

File: lib/tools/apk-utils.js [hunk 21]
```diff
@@ -473,13 +473,79 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
+ * Set the locale name of the device under test and the format of the locale is en-US, for example.
+ * This method call setDeviceLanguageCountry, so, please use setDeviceLanguageCountry as possible.
  *
- * @param {string} locale - The name of the new device country.
+ * @param {string} locale - The name of device country connected with `-`. e.g. en-US.
  */
 apkUtilsMethods.setDeviceLocale = async function (locale) {
-  // this method is only used in API >= 23
-  await this.setDeviceSysLocale(locale);
+  let split_locale = locale.split("-");
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: can we add some input validation here?

Comment (mykola-mokhnach) [reply lib/tools/apk-utils.js]: at least that the string has a dash, so split_locale[0] won't raise an unexpected exception

File: lib/tools/apk-utils.js [hunk 22]
```diff
@@ -473,13 +473,79 @@ apkUtilsMethods.getDeviceLocale = async function () {
 };
 
 /**
- * Set the locale name of the device under test.
+ * Set the locale name of the device under test and the format of the locale is en-US, for example.
+ * This method call setDeviceLanguageCountry, so, please use setDeviceLanguageCountry as possible.
  *
- * @param {string} locale - The name of the new device country.
+ * @param {string} locale - The name of device country connected with `-`. e.g. en-US.
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: -> ...Names of the device language and the country...

File: lib/tools/apk-utils.js [hunk 23]
```diff
@@ -473,13 +473,79 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
+ * Set the locale name of the device under test.
+ *
+ * @param {string} language - Language. format: [a-zA-Z]{2,8}. e.g. en, ja : https://developer.android.com/reference/java/util/Locale.html
+ * @param {string} country - Country. format: [a-zA-Z]{2} | [0-9]{3}. e.g. US, JP : https://developer.android.com/reference/java/util/Locale.html
+ */
+apkUtilsMethods.setDeviceLanguageCountry = async function (language, country) {
+  let hasLanguage = language && _.isString(language);
+  let hasCountry = country && _.isString(country);
+  if (!hasLanguage && !hasCountry) {
+    log.debug(`setDeviceLanguageCountry requires language or country.`);
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: ->log.warn

File: lib/tools/apk-utils.js [hunk 24]
```diff
@@ -473,13 +473,79 @@ apkUtilsMethods.getDeviceLocale = async function () {
...
+    log.debug(`language: '${language}' and country: '${country}'`);
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: -> Got language...

File: lib/tools/apk-utils.js [hunk 25]
```diff
@@ -476,9 +476,15 @@ apkUtilsMethods.getDeviceLocale = async function () {
  * Set the locale name of the device under test and the format of the locale is en-US, for example.
  * This method call setDeviceLanguageCountry, so, please use setDeviceLanguageCountry as possible.
  *
- * @param {string} locale - The name of device country connected with `-`. e.g. en-US.
+ * @param {string} locale - Names of the device language and the country connected with `-`. e.g. en-US.
  */
 apkUtilsMethods.setDeviceLocale = async function (locale) {
+  let validateLocale = new RegExp(/[a-zA-Z0-9]+-[a-zA-Z0-9]+/);
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: this can be module level constant. Also, are digits allowed there?

File: lib/tools/apk-utils.js [hunk 26]
```diff
@@ -544,6 +550,7 @@ apkUtilsMethods.setDeviceLanguageCountry = async function (language, country) {
   }
 
   if (wasSettingChanged) {
+    log.info("To apply locale setting via setting persist.sys.locale command, reboot the target device is required.");
```
Comment (mykola-mokhnach) [inline lib/tools/apk-utils.js]: -> Rebooting the device in order to apply new locale via 'setting persist.sys.locale' command...

**Comment by mykola-mokhnach** [comment]
@KazuCocoa We usually squash commits before merging them to master, so then the storyline looks nicer.
